AmplifyPlugins/DataStore/AWSDataStoreCategoryPlugin/Sync/InitialSync/InitialSyncOperation.swift

```diff
@@ -0,0 +1,180 @@
+            log.error(error: error)
+            return nil
+        }
+
+    }
+
+    private func query(nextToken: String? = nil) {
+        guard !isCancelled else {
+            super.finish()
+            return
+        }
+
+        guard let api = api else {
+            finish(result: .failure(DataStoreError.nilAPIHandle()))
+            return
+        }
+
+        let document = GraphQLSyncQuery(from: modelType,
+                                        nextToken: nextToken,
+                                        lastSync: lastSyncTime)
+
+        let request = GraphQLRequest(document: document.stringValue,
+                                     variables: document.variables,
+                                     responseType: SyncQueryResult.self,
+                                     decodePath: document.decodePath)
+
+        _ = api.query(request: request) { asyncEvent in
+            switch asyncEvent {
+            case .failed(let apiError):
+                // TODO: Retry query on error
+                self.finish(result: .failure(DataStoreError.api(apiError)))
+            case .completed(let graphQLResult):
+                self.handleQueryResults(graphQLResult: graphQLResult)
+            default:
+                break
+            }
+        }
+    }
+
+    /// Disposes of the query results: Stops if error, reconciles results if success, and kick off a new query if there
+    /// is a next token
+    private func handleQueryResults(graphQLResult: Result<SyncQueryResult, GraphQLResponseError<SyncQueryResult>>) {
+        guard !isCancelled else {
+            super.finish()
+            return
+        }
+
+        guard let reconciliationQueue = reconciliationQueue else {
+            finish(result: .failure(DataStoreError.nilReconciliationQueues()))
+            return
+        }
+
+        let syncQueryResult: SyncQueryResult
+        switch graphQLResult {
+        case .failure(let graphQLResponseError):
+            finish(result: .failure(DataStoreError.api(graphQLResponseError)))
+            return
+        case .success(let queryResult):
+            syncQueryResult = queryResult
+        }
+
+        let items = syncQueryResult.items
+        lastSyncTime = syncQueryResult.startedAt
+
+        for item in items {
+            reconciliationQueue.offer(item)
+        }
+
+        if let nextToken = syncQueryResult.nextToken {
+            DispatchQueue.global().async {
+                self.query(nextToken: nextToken)
+            }
+        } else {
+            updateModelSyncMetadata()
+        }
+
+    }
+
+    private func updateModelSyncMetadata() {
+        guard !isCancelled else {
+            super.finish()
+            return
+        }
+
+        guard let storageAdapter = storageAdapter else {
+            finish(result: .failure(DataStoreError.nilStorageAdapter()))
+            return
+        }
+
+        let syncMetadata = ModelSyncMetadata(id: modelType.modelName, lastSync: lastSyncTime)
```

Comment:
i was thinking of just storing syncMetadata over lastSyncTime. it would then update it's lasySyncTime as queries are executed, then storageAdapter.save would simply take the stored syncMetadata rather than create a new one. don't have a strong case on this suggestion, just initial thought that we could operate on the models that get retrieved, updated, and saved to the storageAdapter